x-pack/plugins/apm/server/lib/services/annotations/get_derived_service_annotations.ts

```diff
@@ -0,0 +1,114 @@
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { isNumber } from 'lodash';
+import { Annotation, AnnotationType } from '../../../../common/annotations';
+import { SetupTimeRange, Setup } from '../../helpers/setup_request';
+import { ESFilter } from '../../../../typings/elasticsearch';
+import { rangeFilter } from '../../helpers/range_filter';
+import {
+  PROCESSOR_EVENT,
+  SERVICE_NAME,
+  SERVICE_ENVIRONMENT,
+  SERVICE_VERSION
+} from '../../../../common/elasticsearch_fieldnames';
+
+export async function getDerivedServiceAnnotations({
+  setup,
+  serviceName,
+  environment
+}: {
+  serviceName: string;
+  environment?: string;
+  setup: Setup & SetupTimeRange;
+}) {
+  const { start, end, client, indices } = setup;
+
+  const filter: ESFilter[] = [
+    { term: { [PROCESSOR_EVENT]: 'transaction' } },
+    { range: rangeFilter(start, end) },
+    { term: { [SERVICE_NAME]: serviceName } }
+  ];
+
+  if (environment) {
+    filter.push({ term: { [SERVICE_ENVIRONMENT]: environment } });
+  }
+
+  const versions =
+    (
+      await client.search({
+        index: indices['apm_oss.transactionIndices'],
+        body: {
+          size: 0,
+          track_total_hits: false,
+          query: {
+            bool: {
+              filter
+            }
+          },
+          aggs: {
+            versions: {
+              terms: {
+                field: SERVICE_VERSION
+              }
+            }
+          }
+        }
+      })
+    ).aggregations?.versions.buckets.map(bucket => bucket.key) ?? [];
+
+  if (versions.length > 1) {
+    const annotations = await Promise.all(
+      versions.map(async version => {
+        const response = await client.search({
+          index: indices['apm_oss.transactionIndices'],
+          body: {
+            size: 0,
+            query: {
+              bool: {
+                filter: filter
+                  .filter(esFilter => !Object.keys(esFilter).includes('range'))
+                  .concat({
+                    term: {
+                      [SERVICE_VERSION]: version
+                    }
+                  })
+              }
+            },
+            aggs: {
+              first_seen: {
+                min: {
+                  field: '@timestamp'
+                }
+              }
+            },
+            track_total_hits: false
+          }
+        });
+
+        const firstSeen = response.aggregations?.first_seen.value;
+
+        if (!isNumber(firstSeen)) {
+          throw new Error(
+            'First seen for version was unexpectedly undefined or null.'
+          );
+        }
+
+        if (firstSeen < start || firstSeen > end) {
+          return null;
+        }
+
+        return {
+          type: AnnotationType.VERSION,
+          id: version,
+          time: firstSeen,
```

Comment:
don't we normally call this `timestamp`? or `timestamp.us` to indicate the precision?

Reply:
changed to `@timestamp`